hazelcast/src/main/java/com/hazelcast/instance/impl/DefaultAddressPicker.java

```diff
@@ -381,30 +380,43 @@ AddressDefinition pickMatchingAddress(Collection<InterfaceDefinition> interfaces
                 if (address == null) {
                     continue;
                 }
-                matchingAddress = address;
-
-                if (preferIPv6Addresses) {
-                    // IPv6 address is preferred, return if address is IPv6.
-                    if (inetAddress instanceof Inet6Address) {
-                        return matchingAddress;
-                    }
-                } else if (inetAddress instanceof Inet4Address) {
-                    // No IPv6 address preference, return if address is IPv4.
-                    return matchingAddress;
+                if (getPriority(address, preferIPv6Addresses) > getPriority(matchingAddress, preferIPv6Addresses)) {
+                    matchingAddress = address;
                 }
             }
         }
         // nothing matched to IP version preference, return what we have.
         return matchingAddress;
     }
 
+    @SuppressWarnings("checkstyle:MagicNumber")
```

Comment:
```suggestion
    /**
     * The priority in desc-order:
     * <ol>
     * <li>IPv6-preference
     * <li>global address
     * <li>site-local address
     * <li>link-local address
     * <li>loopback address
     * </ol>
     */
    @SuppressWarnings("checkstyle:MagicNumber")
```

Reply:
I think adding some documentation to explain the intended outcome would help.

Reply:
Good point. Thanks for the suggestion.